pkgdb/src/lock-flake-installable.cc

```diff
@@ -0,0 +1,320 @@
+  {
+    auto outputSpec = std::visit(
+      overloaded {
+        [&]( const nix::ExtendedOutputsSpec::Default & ) -> nix::OutputsSpec
+        {
+          std::set<std::string> outputsToInstall;
+          auto metaOutputsToInstallCursor = cursor->findAlongAttrPath(
+            nix::parseAttrPath( *state, "meta.outputsToInstall" ) );
+          if ( metaOutputsToInstallCursor )
+            {
+              for ( auto output :
+                    ( *metaOutputsToInstallCursor )->getListOfStrings() )
+                {
+                  outputsToInstall.insert( output );
+                }
+            }
+          else if ( ! outputNames.empty() )
+            {
+              outputsToInstall.insert( outputNames.front() );
+            }
+
+          // this seems to be the default in a few nix places
+          // could also dropt this, since reaching this point means
+          // that the package has no outputs?!
+          if ( outputsToInstall.empty() ) { outputsToInstall.insert( "out" ); }
+          return nix::OutputsSpec::Names { std::move( outputsToInstall ) };
+        },
+        [&]( const nix::ExtendedOutputsSpec::Explicit & e ) -> nix::OutputsSpec
+        { return e; },
+      },
+      extendedOutputsSpec.raw() );
+
+    outputsToInstall = std::visit(
+      overloaded {
+        [&]( const nix::OutputsSpec::Names & n ) -> std::set<std::string>
+        { return n; },
+        [&]( const nix::OutputsSpec::All & ) -> std::set<std::string>
+        {
+          std::set<std::string> outputNamesSet;
+          for ( auto output : outputNames ) { outputNamesSet.insert( output ); }
+          return outputNamesSet;
+        } },
+      outputSpec.raw() );
+  }
+
+  // Store the current system
+  std::string system;
+  {
+    system = this->system;
+  }
+
+
+  // Read `name` field - field is impliend by the derivation
+  std::string name;
+  {
+    auto nameCursor
+      = cursor->findAlongAttrPath( nix::parseAttrPath( *state, "name" ) );
+
+    if ( ! nameCursor ) { return EXIT_FAILURE; }
+    name = ( *nameCursor )->getString();
+  }
+
+  // Read `pname` field
+  std::optional<std::string> pname;
+  {
+    auto pnameCursor
+      = cursor->findAlongAttrPath( nix::parseAttrPath( *state, "pname" ) );
+
+    if ( pnameCursor ) { pname = ( *pnameCursor )->getString(); }
+  }
+
+  // Read `version` field
+  std::optional<std::string> version;
+  {
+    auto versionCursor
+      = cursor->findAlongAttrPath( nix::parseAttrPath( *state, "version" ) );
+
+    if ( versionCursor ) { version = ( *versionCursor )->getString(); }
+  }
+
+  // Read `meta.description` field
+  std::optional<std::string> description;
+  {
+    auto descriptionCursor
+      = cursor->findAlongAttrPath( { state->sMeta, state->sDescription } );
+
+    if ( descriptionCursor )
+      {
+        description = ( *descriptionCursor )->getString();
+      }
+  }
+
+  std::optional<std::string> license;
+  {
+    // todo
+  }
```

Comment:
question: what is the current expected type/format for license(s)?
I believe our current lockfile, expects a single string, but we were set out to support multiple licenses where applicable.

Reply:
I don't quite understand - I think it's a single string in nixpkgs and the catalog API. So why would we make it a list? It's plural in the manifest but only because it's a list of licenses to allow in the entire manifest.

Reply:
in nixpkgs its `license` but it may be a list and structured.

```
nix eval nixpkgs#rustc.meta.license --json
[
  {
    "deprecated": false,
    "free": true,
    "fullName": "MIT License",
    "redistributable": true,
    "shortName": "mit",
    "spdxId": "MIT",
    "url": "https://spdx.org/licenses/MIT.html"
  },
  {
    "deprecated": false,
    "free": true,
    "fullName": "Apache License 2.0",
    "redistributable": true,
    "shortName": "asl20",
    "spdxId": "Apache-2.0",
    "url": "https://spdx.org/licenses/Apache-2.0.html"
  }
]
```

I believe in the catalog we extract the `spdxId` ids?

Reply:
Oh lol sometimes it's a list and sometimes it's a single element:
```
> nix eval nixpkgs#hello-unfree.meta.license --json |
jq
{
  "deprecated": false,
  "free": false,
  "fullName": "Unfree",
  "redistributable": false,
  "shortName": "unfree"
}
> nix eval nixpkgs#rustc.meta.license --json | jq
[
  {
    "deprecated": false,
    "free": true,
    "fullName": "MIT License",
    "redistributable": true,
    "shortName": "mit",
    "spdxId": "MIT",
    "url": "https://spdx.org/licenses/MIT.html"
  },
  {
    "deprecated": false,
    "free": true,
    "fullName": "Apache License 2.0",
    "redistributable": true,
    "shortName": "asl20",
    "spdxId": "Apache-2.0",
    "url": "https://spdx.org/licenses/Apache-2.0.html"
  }
]
```

I'm not opposed to storing a list, but if we do I'd want it to be called `licenses`

Reply:
yeah licenses are a bit tricky since the format is not enforced
in nixpkgs these objects come from `lib.licenses`, but in flakes this can be who know what.
From my previous experience with this: `https://github.com/NixOS/nixos-search/blob/main/flake-info/src/data/import.rs#L35`

Reply:
Lol that is a gorgeous type `Option<OneOrMany<StringOrStruct<License>>>`